share/wake/lib/core/json.wake

```diff
@@ -13,161 +13,196 @@
+        def deeper = omap ("{indention}{_}") depth
+        def tabbed = omap ("\n{_}") depth | getOrElse ""
+        def tabbeder = omap ("\n{_}") deeper | getOrElse ""
+        match lhs
+            JString s = '"', fmt.getJSONFormatString s, '"', rhs
+            JInteger i = fmt.getJSONFormatInteger i, rhs
+            JDouble d = fmt.getJSONFormatDouble d, rhs
+            JBoolean True = "true", rhs
+            JBoolean False = "false", rhs
+            JNull = "null", rhs
+            JArray list =
+                def helper value acc = ",{tabbeder}", rec acc deeper value
+                if list.empty then
+                    '[]', rhs
+                else
+                    "[{tabbeder}", foldr helper ("{tabbed}]", rhs) list | tail
+            JObject list =
+                def helper (Pair key value) acc =
+                    ",{tabbeder}", '"', jsonEscape key, ("\":{space}", rec acc deeper value)
+                if list.empty then
+                    '{}', rhs
+                else
+                    "\{{tabbeder}", foldr helper ("{tabbed}\}", rhs) list | tail
+    rec Nil (if indent > 0 then Some "" else None) lhs
+
 
 export def defaultJSONFormat: JSONFormat =
-  def formatDouble d =
-    match (dclass d)
-      DoubleInfinite if d <. 0e0 = "-Infinity"
-      DoubleInfinite = "Infinity"
-      DoubleNaN      = "NaN"
-      _              = dstr d
-  JSONFormat jsonEscape str formatDouble 0
+    def formatDouble d = match (dclass d)
+        DoubleInfinite if d <. 0e0 = "-Infinity"
+        DoubleInfinite = "Infinity"
+        DoubleNaN = "NaN"
+        _ = dstr d
+    JSONFormat jsonEscape str formatDouble 0
+
 
 export def prettyJSONFormat = defaultJSONFormat | setJSONFormatIndent 2
 
+
 export def customFormatJSON fmt body = doFormat fmt body | cat
+
+
 export def formatJSON = customFormatJSON defaultJSONFormat
+
+
 export def prettyJSON = customFormatJSON prettyJSONFormat
 
+
 export def root /| filterFn = jfilter filterFn root
+
+
 export def jfilter filterFn root = match root
-  JArray l = JArray (filter filterFn l)
-  _ = if filterFn root then JArray (root, Nil) else JArray Nil
+    JArray l = JArray (filter filterFn l)
+    _ = if filterFn root then JArray (root, Nil) else JArray Nil
+
 
 export def root /../ filterFn = jfind filterFn root
+
+
 export def jfind filterFn root =
-  def helper node acc = match node
-    JArray l =
-      def tail = foldr helper acc l
-      if filterFn node then node, tail else tail
-    JObject l =
-      def tail = foldr (helper _.getPairSecond _) acc l
-      if filterFn node then node, tail else tail
-    _ =
-      if filterFn node then node, acc else acc
-  helper root Nil | JArray
+    def helper node acc = match node
+        JArray l =
+            def tail = foldr helper acc l
+            if filterFn node then node, tail else tail
+        JObject l =
+            def tail = foldr (helper _.getPairSecond _) acc l
+            if filterFn node then node, tail else tail
```

Comment:
See what I mean with the single line?  The `then` is hard enough to spot, and the only reason the `else` isn't automatically read as just another object name is because `tail` is repeated on both sides (which is even more insidious in this case because `tail` is indeed normally a function, if it wasn't being shadowed here).